### PR TITLE
Implement HTMLOptionElement.index

### LINKED ISSUE #25392: Implement HTMLOptionElement.index
WPT html/semantics/forms/the-option-element/option-index.html tests it, Servo has it commented out in IDL, https://html.spec.whatwg.org/multipage/form-elements.html#concept-option-index defines it (notably for test purposes, the definition is always a number even when the node is detached).

### PR DESCRIPTION
**Pull request by pshaughn**
Made list_of_options public in HTMLSelectElement so HTMLOptionElement can see it, added index-getting to HTMLOptionElement.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #25392

- [X] There are tests for these changes OR

### CODE REVIEW

**Comment by pshaughn** [comment]
Not sure if this is going to change idlharness results
@bors-servo try=wpt

File: components/script/dom/htmloptionelement.rs [hunk 1]
```diff
@@ -93,6 +94,41 @@ impl HTMLOptionElement {
...
         }
     }
+
+    // https://html.spec.whatwg.org/multipage/#concept-option-index
+    fn index(&self) -> i32 {
+        if let Some(parent) = self.upcast::<Node>().GetParentNode() {
+            if let Some(select_parent) = parent.downcast::<HTMLSelectElement>() {
+                // return index in parent select's list of options
+                return self.index_in_select(select_parent);
+            } else if parent.downcast::<HTMLOptGroupElement>().is_some() {
```
Comment (jdm) [inline components/script/dom/htmloptionelement.rs]: `parent.is::<HTMLOptGroupElement>()` is a bit easier to read.

**Review by jdm — CHANGES_REQUESTED** [review]
The rest of the implementation looks sensible!

File: components/script/dom/htmloptionelement.rs (hunk 1 above)
Comment (pshaughn) [reply components/script/dom/htmloptionelement.rs]: Oops, I think I just didn't know is::<T> existed, I will fix that.

**Comment by jdm** [comment]
```
  ▶ Unexpected subtest result in /html/dom/idlharness.https.html?include=HTML.*:
  └ PASS [expected FAIL] HTMLOptionElement interface: document.createElement("option") must inherit property "index" with the proper type
```

**Comment by jdm** [comment]
@bors-servo retry
* #24726

**Comment by jdm** [comment]
@bors-servo retry
* #24726

**Comment by jdm** [comment]
@bors-servo retry
* #24726

**Comment by jdm** [comment]
@bors-servo retry
* #24726

**Pull request merged**